### DISCUSSION TITLE
Searching the SK.Stepper collection for an instance of a Type?

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#786] [Q&A] [answered]
@maluoi First, it seems that there may be multiple instances of the same Type because the documentation for SK.GetStepper<T>() returns the first instance of a Stepper type registered by the SK.AddStepper() method.
IStepper does not contain an ID property (even optionally). The SK Steppers is a private list in SK.cs so we can't iterate over it from our code. GetStepper only uses the Type to locate an instance in the list of Steppers.
There is no way in its current implementation to find an instance of an IStepper from the list, only the first instance. To fully exploit the value of the steppers and the Step() access to the list of Steppers is necessary. Otherwise, we must maintain a completely separate dictionary of Steppers by id and instance of type.

At a minimum, we need to have a method that returns a list of Steppers of a type to allow us to address the properties of the instance or to remove it as needed. Currently I am maintaining a dictionary <string (the ID), string (the type name)>. As long as there is only a single instance of a type then this works fine. It is redundant to maintain a dictionary <string (the ID), IStepper (the activated instance)>

What is really needed is an ID value added to the IStepper and a GetStepper(string ID) that returns the Stepper instance from the Steppers list. This would not change any existing functionality. It would be more in line with the other SK objects that loosely use the ID property e.g UI.Button ID. With this simple change we could use the function provided, such as RemoveStepper().

Note: My argument that nothing changes, is based on the fact that no one has an ID currently and its value is effectively null. I know C# does not like null strings. But the string.IsNullOrEmpty() method is just for such an occurrence. At most, applications upgrading to a new version including an ID property in IStepper would have to at least add a single line to their implementation to set the default value of the ID to string.empty. On the breaking-change scale that is fairly minor for a large gain is function.

**Comment by maluoi** [comment]
It's better to think of `IStepper` as a type of system or manager, and _not_ an object or an item. `IStepper` is designed more for thread safety than performance, so having many instances of an `IStepper` that get added and removed frequently can be pricey.

A better solution would be to create a system that can contain a list you can iterate over.

```csharp
interface IItem {
    void Step();
}

class ItemManager : IStepper {
    public bool Enabled => true;
    public static List<IItem> items = new();

    public void IStepper.Step() {
        foreach(IItem item in items) item.Step();
    }

    public bool IStepper.Initialize() => true;
    public void IStepper.Shutdown  () {}
}
```

Writing something like this of your own gives you additional control over your item's lifetimes, the complexity of adding and removing items, or even the ability to swap them out wholesale for a totally different list. StereoKit can't make assumptions about the thread safety of your `AddStepper` calls, but _you_ can make assumptions about how your `ItemManager` gets used.

And if you're hunting for performance, you can even ditch the `IItem` interface and just use structs!

**Comment by laultman** [comment]
@maluoi I like your suggestions. It is "kind of" back to where I was originally heading but with some very useful suggestions. I don't know why I didn't think of using structs since I certainly know the structure of each of my objects. Sometimes you just get so focused you really don't think.

I had called my version IChain. The change I am about to make is to change to a modular format where a feature is itself a chained set of tasks that have a particular meaning or task to accomplish. I have a customer requirement that says I must switch tasks without leaving the application. My applications are industrial that may require many different tasks related to a single problem domain. Trying to keep the UI manageable for the user while keeping all the data live.

Any other ideas?

**Comment by laultman** [comment, marked as answer]
@maluoi 
UPDATE: I decided to implement my own Stepper tracking service. It has its drawbacks and disadvantages, in that I am forced to use a very few conventions in the coding that may cause some debugging issues. I can overcome most of this by using some project templates specially crafted for StereoKit.

I discovered another issue related to #768 that is mentioned in this thread at the top about the need for an ID on a stepper. I do agree on the comments about the Stepper being thread safe and the implementation of the management of that. It still does not correct the need to locate an instance of a Stepper in SK's list of Steppers. Even though it is allowed to have multiples of the same class Type registered in the stepper list, you cannot locate an instance of one to reset its Enabled property nor can you remove a stepper instance because all you have is a Type. GetStepper method only returns the first instance of the Type. UPDATE: I can now remove, get, or add Steppers. I still can't set the Enabled property.

Seems to me this is likely why you didn't implement the Enabled property to skip the Step method if it is disabled. It will not be trivial.

**Comment by laultman** [comment]
@maluoi a bit of time has passed since I wrote about the issues expressed and the need for an ID. **-Scratch that!** I figured out how to manage all the Stepper registered objects, without the need of an explicit ID. Looking into the code, I found that the Stepper registration contains an ID that is private/internal to the registration. It is created in the AddStepper(). AddStepper() requires an instantiated object, it is not a list of Types to call. What I really needed was a list of Types that could be instantiated at runtime and some management of those.

My solution divides the problem into those "always" living objects and those "intermittent" use objects. The always living are simply created as you would in the AddStepper() and just ignore them from a management standpoint. That leaves those that I care about. My code scans the assemblies in the app disregarding all the well-known namespaces seeking those with an assembly name convention prefix. In the assembly I then select the objects by their Type. By using a Type selection based on an interface, I can be very selective or take a clutch of Types. At runtime then I can invoke their constructors and use code to track the instances. With the instance reference, I can then use the SK methods to remove (unregister) as necessary in my code.

This opens lots of possibilities.
